Mathlib/Analysis/Normed/Affine/AsymptoticCone.lean

```diff
@@ -0,0 +1,428 @@
+/-
+Copyright (c) 2025 Attila Gáspár. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Attila Gáspár
+-/
+
+import Mathlib.Analysis.Convex.Between
+import Mathlib.Analysis.Normed.Module.FiniteDimension
+
+/-!
+# Asymptotic cone of a set
+
+This file defines the asymptotic cone of a set in a normed affine space. The asymptotic cone of a
+set $A$ is usually defined as the set of points $v$ for which there exist sequences $t_n > 0$ and
+$x_n \in A$ such that $t_n \to 0$ and $t_n x_n \to v$. We take a different approach here using
+filters: we define the asymptotic cone of `s` as the set of vectors `v` such that
+`∃ᶠ p in Filter.atTop • 𝓝 v, p ∈ s` holds.
+
+## Main definitions
+
+* `asymptoticNhds`: the filter of neighborhoods at infinity in some direction.
+* `asymptoticCone`: the asymptotic cone of a subset of a normed affine space.
+
+## Main statements
+
+* `Convex.smul_vadd_mem_of_isClosed_of_mem_asymptoticCone`: if `v` is in the asymptotic cone of a
+closed convex set `s`, then every ray of direction `v` starting from `s` is contained in `s`.
+* `Convex.smul_vadd_mem_of_mem_nhds_of_mem_asymptoticCone`: if `v` is in the asymptotic cone of a
+convex set `s`, then every ray of direction `v` starting from the interior of `s` is contained in
+`s`.
+* `isBounded_iff_asymptoticCone_subset_singleton`: in finite dimensions, the asymptotic cone is
+trivial if and only if the set is bounded.
+-/
+
+open scoped Pointwise Topology
+open Filter
+
+section General
+
+variable
+  {k V P : Type*}
+  [NormedField k] [LinearOrder k] [OrderTopology k] [IsStrictOrderedRing k]
+  [NormedAddCommGroup V] [NormedSpace k V] [MetricSpace P] [NormedAddTorsor V P]
+
+variable (k P) in
+/-- In a normed affine space `P` over `k`, `asymptoticNhds k P v` is the filter of neighborhoods at
+infinity in the direction of `v`. In a normed vector space, this is the filter
+`Filter.atTop • 𝓝 v`. To support affine spaces, the actual definition is different and should be
+considered an implementation detail. -/
+@[irreducible]
```

Comment:
If you write it as `irreducible`, it might make sense to add a lemma that unfolds it, and eventually a `mem_asymptoticNhds_iff`.

Reply:
The intention is that only `asymptoticNhds_eq_smul` and `asymptoticNhds_eq_smul_vadd` should be used for unfolding the definition. I could expand the docstring to clarify this. The reason for the different definition is that it cannot depend on some base point in an affine space.